src/broadcast.rs

```diff
@@ -70,61 +69,87 @@
 //! let mut payload: Vec<_> = vec![0; 128];
 //! rng.fill_bytes(&mut payload[..]);
 //!
-//! // Now we can start the algorithm, its input is the payload to be broadcast.
-//! let mut next_message = {
-//!        let proposer = nodes.get_mut(&PROPOSER_ID).unwrap();
-//!        proposer.input(payload.clone()).unwrap();
-//!
-//!        // attach the sender to the resulting message
-//!        proposer.next_message().map(|tm| (PROPOSER_ID, tm))
+//! // Now we can start the algorithm, its input is the payload.
+//! let initial_step = {
+//!     let proposer = nodes.get_mut(&PROPOSER_ID).unwrap();
+//!     proposer.input(payload.clone()).unwrap()
 //! };
 //!
-//! // We can sanity-check that a message is scheduled by the proposer:
-//! assert!(next_message.is_some());
+//! // Initialize the message queue for the simulated network. Annotate the resulting messages with
+//! // the sender ID.
+//! let mut messages: VecDeque<(_, TargetedMessage<_, _>)> = initial_step
+//!     .messages
+//!     .into_iter()
+//!     .map(|tm| (PROPOSER_ID, tm))
+//!     .collect();
+//!
+//! // Initalize the outputs of all nodes with the annotated outputs of the proposer.
+//! let mut outputs: VecDeque<(_, _)> = initial_step
+//!     .output
+//!     .into_iter()
+//!     .map(|v| (PROPOSER_ID, v))
+//!     .collect();
+//!
+//! // We can check that a message is scheduled by the proposer. There should be one as long as the
+//! // number of nodes is greater than 1.
+//! assert!(!messages.is_empty());
 //!
-//! // The network is simulated by passing messages around from node to node.
-//! while let Some((sender, TargetedMessage { target, message })) = next_message {
+//! // The message loop: The network is simulated by passing messages around from node to node.
+//! while let Some((sender, TargetedMessage { target, message })) = messages.pop_front() {
 //!     println!("Message [{:?} -> {:?}]: {:?}", sender, target, message);
 //!
 //!     match target {
 //!         Target::All => {
-//!             let msg = &message;
-//!             nodes.iter_mut()
-//!                  .for_each(|(_, node)| { node.handle_message(&sender, msg.clone())
-//!                                              .expect("could not handle message"); });
-//!         },
-//!         Target::Node(ref dest) => {
-//!             let dest_node = nodes.get_mut(dest).expect("destination node not found");
-//!             dest_node.handle_message(&sender, message)
-//!                      .expect("could not handle message");
-//!         },
+//!             nodes.iter_mut().for_each(|(id, node)| {
+//!                 let step = node
+//!                     .handle_message(&sender, message.clone())
+//!                     .expect("message was handled");
+//!                 // Annotate messages and outputs.
+//!                 messages.extend(step.messages.into_iter().map(|x| (*id, x)));
+//!                 outputs.extend(step.output.into_iter().map(|x| (*id, x)));
```

Comment:
We could verify the output right away, and just take a note which nodes have already output. I.e. just keep a mutable set `finished_nodes` and do something like:

```rust
assert!(step.output.iter().eq(once(payload)));
assert!(finished_nodes.insert(*id));
```